network/src/client.rs

```diff
@@ -139,37 +117,32 @@ macro_rules! define_method {
 }
 
 impl Client {
-    pub fn register_extension(&self, extension: Arc<NetworkExtension>) {
+    pub fn register_extension<T>(&self, extension: Arc<T>)
+    where
+        T: 'static + Sized + NetworkExtension, {
         let name = extension.name();
         let mut extensions = self.extensions.write();
-        if let Some(_) = extensions.insert(name, Arc::clone(&extension)) {
-            let name = extension.name();
-            panic!("Duplicated extension name : {}", name);
-        }
-    }
-
-    pub fn initialize_extension(&self, extension_name: &str) {
-        let extension = {
-            let mut extensions = self.extensions.read();
-            extensions.get(extension_name).map(Arc::clone)
-        };
-        if let Some(extension) = extension {
-            let p2p_channel = self.p2p_channel.clone();
-            let timer_channel = self.timer_channel.clone();
-            let api: Arc<Api> = Arc::new(ClientApi {
-                extension: Arc::downgrade(&extension),
-                p2p_channel,
-                timer_channel,
-            });
-            extension.on_initialize(api);
+        match extensions.entry(name) {
```

Comment:
I cannot find the reason to use `entry` method.
`insert` method gives a more abstraction. Please use it.